### ISSUE TITLE
Question about the render function of inverse depth

### ISSUE THREAD

**Issue by RM-Zhang** [graphdeco-inria/gaussian-splatting#1211]
Hi, thanks for your excellent work.

I noticed that the code computes the weighted average of the inverse depths of all Gaussians to get the rendered inverse depth value.

https://github.com/graphdeco-inria/diff-gaussian-rasterization/blob/9c5c2028f6fbee2be239bc4c9421ff894fe4fbe0/cuda_rasterizer/forward.cu#L374-L375

But from my understanding, it should first compute the weighted average of the depths, and then take the inverse. These two ways are not equivalent, so I’m wondering: is it a bug, or did I misunderstand something?

**Comment by ali-john** [comment]
Hi @RM-Zhang,
Do you have any idea how to use the invdepth if I want to get the actual depth back? Is it simply `1/invdepth` ?  I need this in order to project a pixel back into 3D, but when I do this, I am not getting correct projections. 
Thanks.

**Comment by RM-Zhang** [comment]
Hi @ali-john ,
I replace
``
expected_invdepth += (1 / depths[collected_id[j]]) * alpha * T;
``
with
``
expected_depth += depths[collected_id[j]] * alpha * T;
``
in forward.cu, and return it as the rendered depth. It seems to work well.

Hope it helps!

**Comment by ali-john** [comment]
Hello @RM-Zhang,

Thank you for your response to my previous query. I implemented your suggestions, and also tried recording the distance at which the first Gaussian is hit to use as depth. Both approaches seem to produce very similar results.

However, I’m currently facing another issue and would greatly appreciate your insights.

When rendering an image, I select certain pixels based on specific criteria. I then project these pixels into 3D space, and I expect them to be located close to the Gaussians. However, in practice, the projections appear to be offset. You can see this in [this figure](https://drive.google.com/file/d/13IPAjuAmpdtyRMFWaTW1KorrpnkMA4OZ/view?usp=sharing): the blue points represent the Gaussians, while the light grey points show the projected pixels.

I suspect I might be missing a normalization step or some transformation detail. Below is the code I’m currently using for the 3D projection.
Thank you in advance for your help!


```
def _project_depth(self, u: int, v: int, depth: float, camera) -> np.ndarray:
        if depth <= 0:
          return None
  
        W, H = camera.image_width, camera.image_height
        #  intrinsics 
        fy =  fov2focal(camera.FoVy, camera.image_height)
        fx = fov2focal(camera.FoVx, camera.image_width)
        cx, cy = W/2.0, H/2.0

        # camera space
        x_cam = (u - cx) * depth / fx
        y_cam = -(v - cy) * depth / fy
        z_cam = -depth

        # to homogeneous and world
        cam_coords = torch.tensor([x_cam, y_cam, z_cam, 1.0],
                                dtype=torch.float32,
                                device=camera.data_device)
        cam_to_world = camera.world_view_transform.inverse().to(torch.float32)
        world = cam_to_world @ cam_coords
        return world[:3].cpu().numpy()
```

Details:
Dataset used: ActorsHQ

**Comment by RM-Zhang** [comment]
Hi, @ali-john 

If rendering with Gaussians works correctly, then the issue likely stems from mismatches in coordinate conventions or transformations. I would suggest checking the following:

1. Camera Coordinate System: The world_view_transform in the Camera class is constructed using a column-vector convention (as in OpenGL), whereas most operations in PyTorch follow a row-major convention (similar to OpenCV). This mismatch may lead to incorrect transformations.

2. Depth Units: Make sure the depth values are consistent—are they in meters or millimeters?

Hope this helps!
